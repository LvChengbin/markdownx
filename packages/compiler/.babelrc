{
    "presets" : [ "@babel/preset-react" ],
    "plugins" : [
        [ "import", { "libraryDirectory" : "es", "style" : true } ]
    ]
}
