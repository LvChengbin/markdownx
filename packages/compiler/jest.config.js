module.exports = {
    testMatch : [
        '**/test/**/*.spec.js',
        '**/test/**/*.spec.ts',
        '**/test/**/*.spec.jsx',
        '**/test/**/*.spec.tsx'
    ],
    coverageReporters : [
        'text-summary',
        'text',
        'lcov'
    ],
    collectCoverageFrom : [
        'src/**/*.js'
    ],
    testEnvironment : 'node'
}
