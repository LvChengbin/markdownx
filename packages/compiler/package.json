{
    "name": "@markdownx/compiler",
    "version": "1.0.12",
    "main": "lib/index.js",
    "files": [
        "lib"
    ],
    "author": "LvChengbin <lvchengbin59@gmail.com>",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf lib",
        "build": "yarn clean && tsc -b -v && run replace-internal-path",
        "test": "jest",
        "replace-internal-path": "node scripts/replace-internal-path.js",
        "lint": "eslint {src,test}/**/*.ts --cache --fix --ignore-path .eslintignore",
        "demo": "webpack serve --config webpack.demo.config.js --mode development --host 0.0.0.0 --hot",
        "postpack": "run clean",
        "prepack": "run build"
    },
    "engines": {
        "node": ">=14.15.0"
    },
    "devDependencies": {
        "@emotion/react": "^11.4.0",
        "@emotion/styled": "^11.3.0",
        "@material-ui/core": "next",
        "@material-ui/icons": "next",
        "@material-ui/styles": "next",
        "@testing-library/jest-dom": "^5.13.0",
        "@types/jest": "^26.0.23",
        "@types/marked": "^2.0.3",
        "@types/node": "^16.0.0",
        "@types/react": "^17.0.13",
        "@types/react-dom": "^17.0.8",
        "@typescript-eslint/eslint-plugin": "^4.28.1",
        "@typescript-eslint/parser": "^4.28.1",
        "eslint": "^7.30.0",
        "eslint_d": "^10.1.3",
        "file-loader": "^6.2.0",
        "html-webpack-plugin": "^5.3.2",
        "jest": "^27.0.6",
        "jest-extended": "^0.11.5",
        "react": "^17.0.2",
        "react-dom": "^17.0.2",
        "react-router-dom": "^5.2.0",
        "replace": "^1.2.1",
        "ts-jest": "^27.0.3",
        "ts-loader": "^9.2.3",
        "typescript": "^4.3.5",
        "webpack": "^5.42.0",
        "webpack-cli": "^4.7.2",
        "webpack-dev-server": "^3.11.2"
    },
    "dependencies": {
        "clsx": "^1.1.1",
        "escape-string-regexp": "^4.0.0",
        "marked": "^2.1.3",
        "query-string": "^7.0.1",
        "react-syntax-highlighter": "^15.4.3"
    },
    "peerDependencies": {
        "@material-ui/core": "*",
        "@material-ui/icons": "*",
        "@material-ui/styles": "*",
        "react": "*",
        "react-dom": "*"
    }
}
