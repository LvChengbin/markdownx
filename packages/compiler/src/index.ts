/******************************************************************
 * Copyright (C) 2020 LvChengbin
 *
 * File: src/index.js
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 05/11/2020
 * Description:
 ******************************************************************/

import qs from 'query-string';
import marked from 'marked';
import renderers from './renderers';
import { frags } from './frags';
import increment from './index-counter';
import parser from './parser';
import templates, { ReactComponentTemplateOptions } from './templates';
import generateImports from './generate-imports';
import { Renderer } from './interfaces';

export * from './interfaces';
export * from './frags';

export interface CompileOptions {
    path: string;
    query: { [key: string]: string };
    type?: string;
    template?: ( options: ReactComponentTemplateOptions ) => string;
    prepend?: string;
    renderers?: Record<string, Renderer>;
};

const rendererMethods = [
    'code', 'blockquote', 'html', 'heading', 'hr', 'list', 'listitem',
    'checkbox', 'paragraph', 'table', 'tablerow', 'tablecell',
    'strong', 'em', 'codespan', 'br', 'del', 'link', 'image', 'text'
];

const markedDefaults = { ...marked.defaults };

export default ( options: CompileOptions ): ( source: string ) => string => {

    const mixedRenderers = { ...renderers, ...( options.renderers ?? {} ) };
    const globalImports: string[] = [];
    const query = options.query;
    const importPool: Set<[string, string?]> = new Set();
    const toc: { text: string; level: number }[] = [];

    // @ts-expect-error marked.defaults is readonly but I really need to reset the content here
    marked.defaults = { ...markedDefaults };
    marked.setOptions( {} );

    marked.use( {
        walkTokens( token ) {
            if( token.type === 'code' ) {
                const id = increment();
                const moduleId = `MDX_CODE_MODULE_${id}`;
                const parsed = parser.code.token( token, { moduleId } );
                const { imports, token : parsedToken } = parsed;

                if( parsedToken.invocation ) {
                    imports.forEach( x => {
                        const q = { ...query, pickType : token.type, pickId : id };
                        const request = `${options.path}?${qs.stringify( q )}`;
                        globalImports.push( `import ${x} from "${request}";` );
                    } );

                    frags.set( id + '', {
                        ...options,
                        token : { ...token },
                        id, moduleId,
                        parsedToken,
                        content : parsedToken.text
                    } );

                    Object.assign( token, parsedToken );

                    token.lang += ` {{{${parsedToken.invocation}}}}`;
                }
            } else if( token.type === 'image' ) {
                const id = increment();
                const moduleId = `MDX_IMAGE_MODULE_${id}`;
                const { href } = token;
                const query = href.split( '?' )[ 1 ] ?? '';
                /**
                 * Skip the src if it is able to be treated as a valid URL.
                 */
                if( !href || href.startsWith( '/' ) || href.startsWith( '://' ) || /^https?:\/\//i.test( href ) ) return;
                globalImports.push( `import ${moduleId} from "${token.href}";` );
                token.href = `${moduleId}?${query}`;
            }
        },
        renderer : rendererMethods.reduce( ( acc, cur ) => {
            acc[ cur ] = ( ...args ): string | false => {
                if( cur === 'heading' ) {
                    toc.push( {
                        text : args[ 0 ] as string,
                        level : args[ 1 ] as number
                    } );
                }


                if( mixedRenderers[ cur ] ) {
                    const { imports, exec } = mixedRenderers[ cur ]( options );
                    imports.forEach( x => importPool.add( x ) );
                    return exec( ...args );
                }
                return false;
            };
            return acc;
        }, {} )
    } );

    return ( str: string ): string => {

        let output = marked( str.replace( /(?<=<[^>]+?\s+)((class)(=("|').*?\\4)?)/g, 'className' ).replace( /<!--[\s\S]*?-->/g, '' ) );

        if( mixedRenderers.container ) {
            const { imports, exec } = mixedRenderers.container();
            imports.forEach( x => importPool.add( x ) );
            output = exec( output );
        }

        if( mixedRenderers.toc ) {
            const { imports, exec } = mixedRenderers.toc();
            imports.forEach( x => importPool.add( x ) );
            output = ( exec( toc ) || '' ) + output;
        }

        if( mixedRenderers.root ) {
            const { imports, exec } = mixedRenderers.root();
            imports.forEach( x => importPool.add( x ) );
            output = exec( output );
        }

        return templates[ options.type ?? 'tsx' ]( {
            name : `MDX_COMPONENT_${increment()}`,
            content : output,
            imports : [ ...generateImports( importPool ), ...globalImports ]
        } );
    };
};
