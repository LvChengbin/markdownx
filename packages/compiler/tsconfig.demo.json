{
    "extends" : "../tsconfig.base.json",
    "compilerOptions" : {
        "composite" : true,
        "tsBuildInfoFile" : "./lib/tsconfig.tsbuildinfo"
    },
    "include" : [ "src/**/*", "demo/**/*" ]
}
