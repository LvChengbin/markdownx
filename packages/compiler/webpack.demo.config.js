/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: compiler/webpack.demo.config.js
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 06/15/2021
 * Description:
 ******************************************************************/

const HtmlPlugin = require( 'html-webpack-plugin' );

module.exports = ( env, argv ) => {

    const isProductionMode = argv.mode === 'production';

    return {
        entry : {
            main : './demo/index.tsx'
        },
        output : {
            publicPath : '/'
        },
        resolve : {
            extensions : [ '.ts', '.tsx', '.js', '.jsx' ],
            symlinks : false
        },
        module : {
            rules : [ {
                test : /\.m?js/,
                resolve : {
                    fullySpecified : false
                }
            }, {
                test : /\.tsx?$/,
                use : [ {
                    loader : 'ts-loader',
                    options : {
                        transpileOnly : !isProductionMode,
                        experimentalWatchApi : !isProductionMode,
                        configFile : 'tsconfig.demo.json',
                    }
                } ]
            }, {
                test : /\.(jpe?g|png|gif|svg|webp)$/i,
                loader : 'file-loader'
            } ]
        },
        plugins : [
            new HtmlPlugin( {
                meta : {
                    viewport : 'width=device-width, initial-scale=1, shrink-to-fit=no',
                    charset : 'utf-8'
                },
                templateContent : `
                    <!DOCTYPE html>
                    <html>
                        <head>
                            <title>Documentations: @nextseason/react-ui'</title>
                            <style>html { scroll-behavior : smooth }</style>
                        </head>
                        <body>
                            <div id="root"></div>
                        </body>
                    </html>
                `
            } )
        ],
        devServer : {
            port : 20000,
            historyApiFallback : true,
            hot : true
        }
    };
};
