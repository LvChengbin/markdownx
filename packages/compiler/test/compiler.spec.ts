import compiler from '../src/index';

const str = `## Bubble \
### BubbleBase \
\`\`\`tsx {{{ <$\{MODULE} /> }}} \
interface Props { className?: string; }\
import React from 'react'; \
import BubbleBase from '../component/BubbleBase'; \
export default function SimpleBubble() {\
    return <div class='class-test'> < className='test' BubbleBase>你有什么问题？</BubbleBase> <div>\
}\
\`\`\`
`;

describe( 'compiler', () => {
    describe( 'compiler.', () => {
        it( 'processing "class"', () => {
            const result = compiler( {
                path : '',
                query : {}
            } )( str );
            expect( /className=&#39;class-test&#39;/.test( result ) ).toEqual( true );
        } );

        it( 'processing "className"', () => {
            const result = compiler( {
                path : '',
                query : {}
            } )( str );
            expect( /className\?: string;/.test( result ) ).toEqual( true );
            expect( /classNameName/.test( result ) ).toEqual( false );
        } );
    } );

    describe( 'HTML Tags', () => {
        it( 'Using <iframe /> tag', () => {
            const code = '<iframe style="border: 1px solid rgba(0, 0, 0, 0.1);" width="800" height="450" src="https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2FQ0rPf5bXLBnX8uxOhKXId3%2FUntitled%3Fnode-id%3D0%253A1" allowfullscreen></iframe>';
            const result = compiler( { path : '', query : {} } )( code );

            expect( result ).toMatchSnapshot();
        } );
    } );
} );
