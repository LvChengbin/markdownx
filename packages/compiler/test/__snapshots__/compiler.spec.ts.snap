// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler HTML Tags Using <iframe /> tag 1`] = `
"
import React from 'react';
import HTML from '@@component/HTML';;
import Container from '@@component/Container';;
import Toc from '@@component/Toc';;
import Root from '@@component/Root';
export default function MDX_COMPONENT_2( props: any = {} ): JSX.Element {
    return (
        <div {...props}>
            
            
                <Root>
                <Toc toc={[]} />
            
                <Container>
                <HTML content={\`<iframe style=\\"border: 1px solid rgba(0, 0, 0, 0.1);\\" width=\\"800\\" height=\\"450\\" src=\\"https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2FQ0rPf5bXLBnX8uxOhKXId3%2FUntitled%3Fnode-id%3D0%253A1\\" allowfullscreen></iframe>\`} />
            </Container>
            </Root>
            
        </div>
    );
}
"
`;
